suites/apis/centralizedAuth.js

```diff
@@ -736,28 +765,131 @@ Scenario('Client with user token WITH permission CAN create signed URL for recor
+    console.log('Use user0 to create signed URL for file under `/gen3` with authorized consent code (multiple policies)')
+    const signedUrlRes = await fence.do.createSignedUrlForUser(
+      indexed_files.gen3HmbResearchFile.did, users.user0.accessTokenHeader);
+
+    let fileContents = await fence.do.getFileFromSignedUrlRes(
+      signedUrlRes);
+
+    chai.expect(fileContents,
+      'User with access could NOT create signed urls and read file for records in ' +
+      'authorized namespace with authorized consent code'
+    ).to.equal(fence.props.awsBucketInfo.cdis_presigned_url_test.testdata);
+});
 
 /*
    - Do the same as above but in a single policy
 */
+Scenario('Test create signed URL for file in authorized namespace with authorized consent code (single policies) @centralizedAuth',
+  async (fence, indexd, users, files) => {
+    // dcf-integration-test-1@planx-pla.net (user1)
+    //     - gen3-hmb-researcher
+    //
+    // gen3HmbResearchFile in `/gen3` namespace but also requires `/consents/HMB`
+    console.log('Use user1 to create signed URL for file under `/gen3` with authorized consent code (single policy)')
+    const signedUrlRes = await fence.do.createSignedUrlForUser(
+      indexed_files.gen3HmbResearchFile.did, users.user1.accessTokenHeader);
+
+    let fileContents = await fence.do.getFileFromSignedUrlRes(
+      signedUrlRes);
+
+    chai.expect(fileContents,
+      'User with access could NOT create signed urls and read file for records in ' +
+      'authorized namespace with authorized consent code'
+    ).to.equal(fence.props.awsBucketInfo.cdis_presigned_url_test.testdata);
+});
 
 
 /*
    - Give user access to program/proj but don't give consent code, make sure they cannot access
      the data
 */
+Scenario('Test cannot create signed URL for file in authorized namespace with UNauthorized consent code @centralizedAuth',
+  async (fence, indexd, users, files) => {
+    // cdis.autotest@gmail.com (mainAcct)
+    //     - abc-admin
+    //
+    // abcHmbResearchFile in `/abc` namespace but also requires `/consents/HMB`
+    console.log('Use mainAcct to create signed URL for file under `/abc` with UNauthorized consent code')
+    const signedUrlRes = await fence.do.createSignedUrlForUser(
+      indexed_files.abcHmbResearchFile.did, users.mainAcct.accessTokenHeader);
+
+    let fileContents = await fence.do.getFileFromSignedUrlRes(
+      signedUrlRes);
+
+    chai.expect(fileContents,
+      'User WITHOUT access COULD create signed urls and read file for records in ' +
+      'authorized namespace with UNauthorized consent code'
+    ).to.not.equal(fence.props.awsBucketInfo.cdis_presigned_url_test.testdata);
+});
 
 
 /*
    - Give user access to program/proj and HMB policy, make sure they CAN access
      files with GRU
-*/
\ No newline at end of file
+*/
+Scenario('Test create signed URL for file in authorized namespace with IMPLIED authorized consent code (based on DUO hierarchy) @centralizedAuth',
```

Comment:
i don't understand the point of this test, to me it looks like it tests the same thing as the first test (`Test create signed URL for file in authorized namespace with authorized consent code (multiple policies`)

`HMB ` and `GRU` are both in policy `hmb-researcher` so both tests check that the user has access to "any" code that's in the policy

can you explain? :D

Reply:
yeeah, so the way this is going to appear to work to end-users is you give a user a single policy for what they've consented to and there is an _implied_ hierarchy for the codes. I'll send you links to some diagrams :confetti_ball: on slack.

basically, by having access to HMB, user should also have access to GRU. so this functionality may be somewhat duplicated in other tests that have multiple policies, but this is a specific test case for the consent codes themselves.

I guess it's all configuration, so maybe we don't really _need_ this test :man_shrugging: but it tests the client user case explicitly better than the first test.

I leaned towards writing it to be more explicit about the expectation of the consent codes but I guess I could be convinced to remove it

Reply:
✔️